### PR TITLE
Move esptool from user-venv dependency to vendor it with Mu.

### PR DESCRIPTION
**Pull request by carlosperate**
As discussed in:
- https://github.com/mu-editor/mu/issues/2120#issuecomment-1309468274

Only a single script was needed from the entire esptool package, and the only dependency needed for that script is pyserial, which is already included with Mu.
So, vendoring esptool means we don't need to pip download, package, and install its other dependencies, which have caused trouble in the past (specially cryptography when it moved to be a rust package, as discussed in https://github.com/mu-editor/mu/issues/1952).

This changes the "category" of this dependency, from a wheel packaged with Mu to install in the user venv, to be included as a Mu app runtime dependency. As esptool is used in the Mu settings, is probably closer to a "Mu runtime" dependency, than a "user running their code in a venv" dependency.

Apart from that, esptool and some of its dependencies don't have wheels in PyPI.
To be able to use the `--platform` flag with `pip download` it needs all dependencies to have wheels, so removing this from the user wheels helps with that as well.

### CODE REVIEW

**Comment by carlosperate** [comment]
I haven't tested this properly yet, but it is added here for visibility and review.

File: mu/interface/dialogs.py [hunk 1]
```diff
@@ -460,17 +454,17 @@ def update_firmware(self):
...
+            sys.executable, esptool, device.port
         )
 
         if self.device_type.currentText() == "ESP32":
             write_command = (
                 '"{}" "{}" --chip esp32 --port {} --baud {} '
                 'write_flash -z 0x1000 "{}"'
             ).format(
-                venv.interpreter,
+                sys.executable,
```
Comment (ntoll) [inline mu/interface/dialogs.py:467]: Why the change from `venv.interpreter` to `sys.executable`..?

File: tests/interface/test_dialogs.py [hunk 2]
```diff
@@ -178,11 +178,7 @@ def microbit():
     return device
 
 
-@mock.patch(
-    "mu.interface.dialogs.ESPFirmwareFlasherWidget.esptool_is_installed",
-    return_value=True,
-)
-def test_ESPFirmwareFlasherWidget_setup(esptool_is_installed, microbit):
+def test_ESPFirmwareFlasherWidget_setup(microbit):
```
Comment (ntoll) [inline tests/interface/test_dialogs.py:181]: Nice to see mocks are gone.

**Review by ntoll — APPROVED** [review]
I wonder if we can exclude `esptool.py` from LGTM's analysis too.

File: mu/interface/dialogs.py (hunk 1 above)
Comment (carlosperate) [reply mu/interface/dialogs.py:467]: esptool used to be run from the user venv, which would have `pyserial` installed as part of installing `esptool`.
Now as we vendor it inside the Mu source code we need to execute it in an environemnt that has `pyserial` installed (pyserial is no longer installed in the user venv), so we now have to use `sys.executable` which points to the python executable+environment that is running Mu, which will have `pyserial`.

File: tests/interface/test_dialogs.py (hunk 2 above)
Comment (carlosperate) [reply tests/interface/test_dialogs.py:181]: Yeah, `esptool_is_installed()` wasn't really used anywhere, might have been accidentally left after a refactor at some point.

**Comment by carlosperate** [comment]
Tested this in a couple of ESP8266 devices and works as expected.

File: mu/interface/dialogs.py (hunk 1 above)
Comment (ntoll) [reply mu/interface/dialogs.py:467]: Of course. Thank you for the clarification.

**Comment by carlosperate** [comment]
Since Semmle was adquire by GitHub and LGTM will be shut down in December, we might as well move to the current way to run it:
- https://github.com/mu-editor/mu/pull/2347

And how to exclude files:
https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning#specifying-directories-to-scan

I'll merge the PR here and we can merge it together to the main branch.

**Comment by carlosperate** [comment]
Okay, it's now configured to ignore the esptool script and the warnings went down to 27:
image

I think this PR is ready to be merged 🎉

**Comment by carlosperate** [comment]
@ntoll As now we are using CodeQL directly from GitHub I guess we should deactivate LGTM, which I assume you'd have to do from their website?

**Comment by ntoll** [comment]
@carlosperate yeah... lemme try to remember how to do that... :rofl:

**Comment by carlosperate** [comment]
Great, thanks! I'll merge this as done then 👍
